src/core/server/saved_objects/export/collect_exported_objects.ts

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type { SavedObject } from '../../../types';
+import type { KibanaRequest } from '../../http';
+import { SavedObjectsClientContract } from '../types';
+import type { SavedObjectsExportTransform } from './types';
+import { applyExportTransforms } from './apply_export_transforms';
+
+interface CollectExportedObjectOptions {
+  savedObjectsClient: SavedObjectsClientContract;
+  objects: SavedObject[];
+  /** flag to also include all related saved objects in the export stream. */
+  includeReferences: boolean;
+  /** optional namespace to override the namespace used by the savedObjectsClient. */
+  namespace?: string;
+  /** The http request initiating the export. */
+  request: KibanaRequest;
+  /** export transform per type */
+  exportTransforms: Record<string, SavedObjectsExportTransform>;
+}
+
+interface CollectExportedObjectResult {
+  objects: SavedObject[];
+  missingRefs: CollectedReference[];
+}
+
+export const collectExportedObjects = async ({
+  objects,
+  includeReferences,
+  namespace,
+  request,
+  exportTransforms,
+  savedObjectsClient,
+}: CollectExportedObjectOptions): Promise<CollectExportedObjectResult> => {
+  const collectedObjects: SavedObject[] = [];
+  const collectedMissingRefs: CollectedReference[] = [];
+  const alreadyProcessed: Set<string> = new Set();
+
+  let currentObjects = objects;
+  do {
+    const transformed = await applyTransforms({
```

Comment:
Do we need `applyTransforms` function?
```js
transformed.filter((object) => !alreadyProcessed.has(objKey(object)));
```
It's not transformation logic. maybe, we should inline it in `collectExportedObjects`?

Reply:
Yea, that make sense. Initially I had more logic inside `applyTransforms`, and forget to inline when I moved it.